### PR TITLE
API release automation with go script

### PR DESCRIPTION
**Pull request by Joerger**
This PR is the successor to https://github.com/gravitational/teleport/pull/7598. It is functionally the same to the previous version except that it now uses a go program instead of bash scripting. The initial bash script version was veto'd because a go program will be more readable, maintainable, and less bug prone, especially since it can use `golang.org/x/mod/modfile`, `golang.org/x/tools/go/ast/astutil`, and `golang.org/x/tools/go/packages`.

Note: The go program is not currently to go backwards from v2+ to v0/v1.

### Description:
As a part of the [API RFD](https://github.com/gravitational/teleport/blob/master/rfd/0010-api.md#versioning)...

This PR adds the `update-api-module-path` make target to update the api module path to the current major version.e.g. `gravitational/teleport/api` -> `gravitational/teleport/api/v7`.

The program only makes changes if:
    1. the major version of the API changes
    2. no version suffixes are present `dev|alpha|beta|rc`
      - This way, the import path is only updated on an actual release, which avoids backport and merge-conflict issues during the pre-release rush.

`update-api-module-path` runs at the end of `make version`, so the api version should be automatically updated with the current release processes in place.

Other changes:
 - since `protoc` does not understand go modules, and reads the version suffix as a directory, `protoc` now reads from `vendor/.../teleport/api/vX`.
   - This make it so `.proto` can be updated with the new api module path, and `pb.go` will be properly generated.

Example of changes made when you update the version and run `make version` - https://github.com/gravitational/teleport/pull/8673

### CODE REVIEW

File: build.assets/Dockerfile [hunk 1]
```diff
@@ -125,7 +125,9 @@ RUN (git clone https://github.com/gogo/protobuf.git ${GOPATH}/src/github.com/gog
      git reset --hard ${GOGO_PROTO_TAG} && \
      make install)
 
-ENV PROTO_INCLUDE "/usr/local/include":"${GOPATH}/src":"${GOPATH}/src/github.com/gogo/protobuf/protobuf":"${GOGOPROTO_ROOT}":"${GOGOPROTO_ROOT}/protobuf":"${GOPATH}/src/github.com/gravitational/teleport/lib/services"
+# Note: protoc reads the proto files from /vendor/.../teleport/api/vX rather than
+# /api because protoc does not understand go modules, and reads vX as a directory.
+ENV PROTO_INCLUDE "/usr/local/include":"${GOPATH}/src/github.com/gravitational/teleport/vendor":"${GOPATH}/src/github.com/gogo/protobuf/protobuf":"${GOGOPROTO_ROOT}":"${GOGOPROTO_ROOT}/protobuf"
```
Comment (zmb3) [inline build.assets/Dockerfile:130]: It would be nice to not assume `GOPATH` is set.

File: build.assets/update-api-module-path/main.go [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+)
+
+func init() {
+	utils.InitLogger(utils.LoggingForCLI, log.DebugLevel)
+}
+
+func main() {
+	// the api mod path should only be updated on releases, check for non-release suffixes
+	if strings.Contains(api.Version, "-") {
+		exitWithMessage("the current api version is not a release, continue without updating")
```
Comment (zmb3) [inline build.assets/update-api-module-path/main.go:42]:
```suggestion
		exitWithMessage("the current API version (%v) is not a release, continue without updating", api.Version)
```

File: build.assets/update-api-module-path/main.go [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+					rewritten = true
+				}
+			}
+		}
+		if !rewritten {
+			continue
+		}
+
+		goFileName := p.Fset.File(syn.Pos()).Name()
+		f, err := os.Create(goFileName)
```
Comment (zmb3) [inline build.assets/update-api-module-path/main.go:131]: Why would we ever want to create a new file?

File: build.assets/update-api-module-path/main.go [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+	return filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if strings.HasSuffix(d.Name(), fileExtension) {
+			data, err := ioutil.ReadFile(path)
+			if err != nil {
+				return err
+			}
+			fileString := strings.ReplaceAll(string(data), oldModPath, newModPath)
```
Comment (zmb3) [inline build.assets/update-api-module-path/main.go:199]: No need to go from `[]byte` to `string` and back again. Just use `bytes.ReplaceAll` instead.

File: build.assets/update-api-module-path/main.go [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+}
+
+// getModImportPath gets the module's currently set path/name
+func getModImportPath(dir string) (string, error) {
+	modFile, err := getModFile(dir)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	if modFile.Module.Mod.Path == "" {
+		return "", trace.Errorf("could not get mod path")
```
Comment (zmb3) [inline build.assets/update-api-module-path/main.go:215]: Do you need `Errorf` here? I don't see any formatting directives.

**Review by zmb3 — COMMENTED** [review]
I missed the discussion on the RFD, but I'm a little concerned that we're going against the general [advice](https://golang.org/doc/modules/version-numbers#major-version):

> A module developer should increment this number past v1 only when necessary because the version upgrade represents significant disruption for developers whose code uses function in the upgraded module. This disruption includes backward-incompatible changes to the public API, as well as the need for developers using the module to update the package path wherever they import packages from the module.

File: Makefile [hunk 6]
```diff
@@ -601,6 +601,22 @@ version: $(VERSRC)
...
 	VERSION=$(VERSION) $(MAKE) -f version.mk setver
+	# Update api module path, but don't fail on error.
+	$(MAKE) update-api-module-path || true
+
+# This rule updates the api module path to be in sync with the current api release version.
+# e.g. github.com/gravitational/teleport/api/vX -> github.com/gravitational/teleport/api/vY
+#
+# It will immediately fail if:
+#  1. A suffix is present in the version - e.g. "-alpha"
+#  2. The version suffix hasn't changed - e.g. 7.0.0 -> 7.1.0
```
[the remark this answers is not in the document]
Comment (Joerger) [reply Makefile:612]: I'll update it to `e.g. 7.0.0 -> 7.1.0 (both "/v7") or v0.0.0 -> 1.0.0 (no suffix)`, does that make more sense?

[the remark this answers is not in the document]
Comment (zmb3) [reply Makefile:612]: How about we say "if the major version hasn't changed"

File: build.assets/Dockerfile (hunk 1 above)
Comment (Joerger) [reply build.assets/Dockerfile:130]: It's set within the Dockerfile so it should be fine as is.

**Comment by Joerger** [comment]
> I missed the discussion on the RFD, but I'm a little concerned that we're going against the general [advice](https://golang.org/doc/modules/version-numbers#major-version):
> 
> > A module developer should increment this number past v1 only when necessary because the version upgrade represents significant disruption for developers whose code uses function in the upgraded module. This disruption includes backward-incompatible changes to the public API, as well as the need for developers using the module to update the package path wherever they import packages from the module.

I'm with you. I definitely wanted to avoid it, but it makes things a lot simpler for customers and developers to have Teleport client and server version aligned. Our [version compatibility rules](https://github.com/gravitational/teleport/blob/master/rfd/0012-teleport-versioning.md#compatibility) are already hard for customers and developers to follow at times, so introducing a completely new versioning guideline for the API will increase the room for errors/issues to occur. Ultimately the pros and cons were weighed and this was the consensus.

It's definitely not too late to reconvene and see if we can make another approach. I still think that [Release Option 2](https://github.com/gravitational/teleport/blob/master/rfd/0010-api.md#release-option-2---release-the-api-separately-from-teleport) from the RFD would give a good in between.

However, the API module has been complete but has yet to be released so I'm concerned about pushing it back longer. [Teleport plugins](https://github.com/gravitational/teleport-plugins/blob/a50b0dd773e20221e7861fa4201114b488420686/go.mod#L19) has to import specific commits which is hard to follow and maintain. This is an even bigger problem for customers who want to import a specific version of the API and have no feasible way to do so without our help.

File: Makefile [hunk 7]
```diff
@@ -867,17 +883,25 @@ init-submodules-e: init-webapps-submodules-e
...
+
+# When teleport vendors its dependencies, Go also vendors the local api sub module. To get
+# around this issue, we replace the vendored api package with a symlink to the
+# local module. The symlink should be in vendor/.../api or vendor/.../api/vX if X >= 2.
+.PHONY: vendor-api
+vendor-api:
+	$(eval MOD_PATH=$(shell head -1 api/go.mod | awk '{print $$2;}'))
+	rm -rf vendor/github.com/gravitational/teleport/api
+	mkdir -p vendor/$(shell dirname $(MOD_PATH))
+	ln -rs $(shell readlink -f api) vendor/$(MOD_PATH)
```
[the remark this answers is not in the document]
Comment (zmb3) [reply Makefile]: `readlink -f` only works on GNU/Linux, not the BSD variant of `readlink` that macOS ships with.

[the remark this answers is not in the document]
Comment (Joerger) [reply Makefile]: It's no longer using readlink

```
.PHONY: vendor-api
vendor-api: MOD_PATH := $(shell head -1 api/go.mod | awk '{print $$2;}')
vendor-api:
	rm -rf vendor/github.com/gravitational/teleport/api
	mkdir -p vendor/$(shell dirname $(MOD_PATH))
	ln -rs ./api vendor/$(MOD_PATH
```

**Review by zmb3 — CHANGES_REQUESTED** [review]
Sorry it took so long to get back to this.

Some final thoughts - it seems if this tool fails at any point, it doesn't attempt to undo any of the actions it took and leave the system in the last known good state.

If we expect developers to run this tooling on their dev machines, and the tool makes changes that cant be quickly reverted with git, then maybe it makes sense to expand the error handling a bit.

Lastly, some tests that exercise this functionality would help make it easier to understand what this tool does and what things should look like before and after a run.

**Review by zmb3 — APPROVED** [review]

**Review by tcsc — APPROVED** [review]

**Review by atburke — APPROVED** [review]

**Review by russjones — APPROVED** [review]
Bot.
